targets/TARGET_NUVOTON/TARGET_M480/i2c_api.c

```diff
@@ -0,0 +1,1004 @@
+        uint32_t status = I2C_GET_STATUS(i2c_base);
+
+        switch (status) {
+        case 0x08:  // Start
+        case 0x10:  // Master Repeat Start
+            if (i2c_ctl & I2C_CTL0_STA_Msk) {
+                return 0;
+            } else {
+                break;
+            }
+        case 0xF8:  // Bus Released
+            if ((i2c_ctl & (I2C_CTL0_STA_Msk | I2C_CTL0_STO_Msk)) == I2C_CTL0_STO_Msk) {
+                return 0;
+            } else {
+                break;
+            }
+        }
+#endif
+        I2C_SET_CONTROL_REG(i2c_base, i2c_ctl);
+        if (sync && i2c_poll_status_timeout(obj, i2c_is_trsn_done, NU_I2C_TIMEOUT_STAT_INT)) {
+            err = I2C_ERROR_BUS_BUSY;
+#if NU_I2C_DEBUG
+            MY_I2C_2 = obj->i2c;
+            while (1);
+#endif
+        }
+    }
+
+    i2c_enable_int(obj);
+
+    return err;
+}
+
+static int i2c_poll_status_timeout(i2c_t *obj, int (*is_status)(i2c_t *obj), uint32_t timeout)
+{
+    uint32_t t1, t2, elapsed = 0;
+    int status_assert = 0;
+
+    t1 = us_ticker_read();
+    while (1) {
+        status_assert = is_status(obj);
+        if (status_assert) {
+            break;
+        }
+
+        t2 = us_ticker_read();
+        elapsed = (t2 > t1) ? (t2 - t1) : ((uint64_t) t2 + 0xFFFFFFFF - t1 + 1);
+        if (elapsed >= timeout) {
+#if NU_I2C_DEBUG
+            MY_I2C_T1 = t1;
+            MY_I2C_T2 = t2;
+            MY_I2C_ELAPSED = elapsed;
+            MY_I2C_TIMEOUT = timeout;
+            MY_I2C_2 = obj->i2c;
+            while (1);
+#endif
+            break;
+        }
+    }
+
+    return (elapsed >= timeout);
+}
+
+static int i2c_poll_tran_heatbeat_timeout(i2c_t *obj, uint32_t timeout)
+{
+    uint32_t t1, t2, elapsed = 0;
+    int tran_started;
+    char *tran_pos = NULL;
+    char *tran_pos2 = NULL;
+
+    i2c_disable_int(obj);
+    tran_pos = obj->i2c.tran_pos;
+    i2c_enable_int(obj);
+    t1 = us_ticker_read();
+    while (1) {
+        i2c_disable_int(obj);
+        tran_started = i2c_is_tran_started(obj);
+        i2c_enable_int(obj);
+        if (! tran_started) {    // Transfer completed or stopped
+            break;
+        }
+
+        i2c_disable_int(obj);
+        tran_pos2 = obj->i2c.tran_pos;
+        i2c_enable_int(obj);
+        t2 = us_ticker_read();
+        if (tran_pos2 != tran_pos) {    // Transfer on-going
+            t1 = t2;
+            tran_pos = tran_pos2;
+            continue;
+        }
+
+        elapsed = (t2 > t1) ? (t2 - t1) : ((uint64_t) t2 + 0xFFFFFFFF - t1 + 1);
+        if (elapsed >= timeout) {   // Transfer idle
+#if NU_I2C_DEBUG
+            MY_I2C = obj->i2c;
+            MY_I2C_T1 = t1;
+            MY_I2C_T2 = t2;
+            MY_I2C_ELAPSED = elapsed;
+            MY_I2C_TIMEOUT = timeout;
+            MY_I2C_2 = obj->i2c;
+            while (1);
+#endif
+            break;
+        }
+    }
+
+    return (elapsed >= timeout);
+}
+
+#if 0
```

Comment:
@0xc0170 Please refer to https://github.com/ARMmbed/mbed-os/pull/4608/commits/f09ce861c16e474823d2fe7033c33557ed86264a for the update.